consensus/core/src/broadcaster.rs

```diff
@@ -0,0 +1,227 @@
+
+use std::{
+    cmp::{max, min},
+    sync::Arc,
+    time::Duration,
+};
+
+use consensus_config::AuthorityIndex;
+use futures::{stream::FuturesUnordered, StreamExt as _};
+use tokio::{
+    sync::broadcast,
+    task::JoinSet,
+    time::{error::Elapsed, sleep_until, timeout, Instant},
+};
+use tracing::{trace, warn};
+
+use crate::{
+    block::VerifiedBlock, context::Context, core::CoreSignalsReceivers, error::ConsensusResult,
+    network::NetworkClient,
+};
+
+/// Number of Blocks that can be inflight sending to a peer.
+const BROADCAST_CONCURRENCY: usize = 10;
+
+/// Broadcaster sends newly created blocks to each peer over the network.
+///
+/// For a peer that lags behind or is disconnected, blocks are buffered and retried until
+/// a limit is reached, then old blocks will get dropped from the buffer.
+pub(crate) struct Broadcaster {
+    // Background tasks sending blocks to peers.
+    senders: JoinSet<()>,
+}
+
+impl Broadcaster {
+    pub(crate) fn new<C: NetworkClient>(
+        context: Arc<Context>,
+        network_client: Arc<C>,
+        signals_receiver: &CoreSignalsReceivers,
+    ) -> Self {
+        // Initialize sender tasks.
+        let mut senders = JoinSet::new();
+        for (index, _authority) in context.committee.authorities() {
+            // Skip sending Block to self.
+            if index == context.own_index {
+                continue;
+            }
+            senders.spawn(Self::push_blocks(
+                context.clone(),
+                network_client.clone(),
+                signals_receiver.block_broadcast_receiver(),
+                index,
+            ));
+        }
+        Self { senders }
+    }
+
+    pub(crate) fn stop(&mut self) {
+        // Intentionally not waiting for senders to exit, to speed up shutdown.
+        self.senders.abort_all();
+    }
+
+    /// Runs a loop that continously pushes new blocks received from the rx_block_broadcast
+    /// channel to the target peer.
+    ///
+    /// The loop does not exit until the validator is shutting down.
+    async fn push_blocks<C: NetworkClient>(
+        context: Arc<Context>,
+        network_client: Arc<C>,
+        mut rx_block_broadcast: broadcast::Receiver<VerifiedBlock>,
+        peer: AuthorityIndex,
+    ) {
+        let peer_hostname = context.committee.authority(peer).hostname.clone();
+
+        // Use a simple exponential-decay RTT estimator to adjust the timeout for each block sent.
+        // The estimation logic will be removed once the underlying transport switches to use
+        // streaming and the streaming implementation can be relied upon for retries.
+        const RTT_ESTIMATE_DECAY: f64 = 0.95;
+        const TIMEOUT_THRESHOLD_MULTIPLIER: f64 = 2.0;
+        const TIMEOUT_RTT_INCREMENT_FACTOR: f64 = 1.6;
+        let mut rtt_estimate = Duration::from_millis(200);
+
+        let mut requests = FuturesUnordered::new();
+
+        async fn send_block<C: NetworkClient>(
```

Comment:
Perhaps pacing could avoid bursty traffic here.

Reply:
Would use Arc to reference one copy of the result of each block.serialized() amongst all senders, rather than passing block, work here?

Reply:
The creation of blocks are already rate limited (because each block needs 2f+1 parents from the previous round). There should not be a burst of valid blocks getting created.